### PR TITLE
Replace constraint tests

### PR DESCRIPTION
**Pull request by p-snft**
Replaces lp diff constraint tests by granular unit tests for functionality.

Reasoning:
* The LP files depend on Pyomo internals, thus they change when Pyomo changes things.
* They include tests for stuff from multiple parts of the code. For example, if I change the constraint formulation for the Bus, every constraint test needs to be rewritten, not only the one for the Bus.
* They are hard to understand (for many) and failed tests are hard to fix. The latter is particularly because of the undefined order of lines, so that just looking at a diff will not tell the difference: Constraints might be in a different order but still the same.

I would suggest to have unit tests that
* will be using Pyomo but the results will be independent from the back-end,
* is easy to understand, and
* even internal constraint formulations can be changed without breaking the test, so it really tests for the specified functionally not the implementation.

### CODE REVIEW

**Comment by p-snft** [comment]
In the process, I found out that we actually have some quadratic constraints.

https://github.com/oemof/oemof-solph/blob/c2d44e4a1a3baa93e738e8cc1551bfbde36e1494/src/oemof/solph/flows/_non_convex_flow_block.py#L749-L750

**Comment by p-snft** [comment]
The tests added in [896f73b](https://github.com/oemof/oemof-solph/pull/1080/commits/896f73bda4469f8b1f4c8042c1f7faaf02692e1e) reveal an inconsistency: While shutting down after the optimisation period is not counted, starting up before the optimisation period seems to be counted. In `test_maximum_startups`, I'd expect the flow to be active in the first and the last time step. Only the startup before the last time step would count.

**Comment by p-snft** [comment]
Can I delete the automatic discounting in

https://github.com/oemof/oemof-solph/blob/c2d44e4a1a3baa93e738e8cc1551bfbde36e1494/src/oemof/solph/flows/_non_convex_flow_block.py#L404-L413

and simplify the code this way? The `activity_costs` are a time series anyway, so discounting could easily be considered externally. (@jokochems)

**Comment by p-snft** [comment]
As I did not want to write a test assuring wrong results, I merged `fix/invest_storage_content_revised`. So, this one now also fixes #1030.

**Comment by p-snft** [comment]
The remaining missing test coverage of the `_generic_storage.py` is related to multi-period code which is mostly duplicated code from single-period models, but also partly broken (see https://github.com/oemof/oemof-solph/pull/1039/commits/159731896e127b5f1943394810c2b0c013cc94bc).

**Comment by p-snft** [comment]
Condacy complains because I have slightly changed the signature of a complex function: It says that I introduced a complex function. (In fact, I slightly reduced the complexity by the change.)

**Comment by jokochems** [comment]
> Can I delete the automatic discounting in
> 
> https://github.com/oemof/oemof-solph/blob/c2d44e4a1a3baa93e738e8cc1551bfbde36e1494/src/oemof/solph/flows/_non_convex_flow_block.py#L404-L413
> 
> and simplify the code this way? The `activity_costs` are a time series anyway, so discounting could easily be considered externally. (@jokochems)
> 
> PS: Of course, the same is also valid for `inactivity_costs`, `startup_costs`, and `shutdown_costs`.

I agree that it makes sense to exclude discounting. In general, this should be harmonized across all cost types, but I think since this is a rather rare, probably unused use case here, I think it is safe to remove it here.

**Review by jokochems — APPROVED** [review]
Dear @p-snft,

thanks for this pull request and the very good effort that I highly appreciate. I left a lot of comments, but most of them are rather editorial remarks and/or questions of understanding that should be easy to address. I also like the new test framework and the examples you created, esp. concerning the non convex flow constraints that were rather untested before (though officially covered). Thus, I am happy to approve the pull request.

Nonetheless, I'd like to raise two points. I don't see them as critical, but rather want to raise awareness:
1. It seems to me that not yet all of the constraints are tested one by one as they actually should. I didn't find an explicit test for bus balance constraints or the converter efficiencies for instance. These were at least implicity contained in the tests before, because at least I also checked these when comparing the LP files.
2. Some experimental components did have testing before, but now they don't have it anymore.

I think it is legitimate to proceed since I do not want to interfere with your development agenda, but rather raise awareness for this issues. I think both aspects can be covered in dedicated pull requests.

Once again, many thanks for the great effort. Besides vastly improving maintainability, in my view also a further improvement in code quality.

File: src/oemof/solph/_models.py [hunk 1]
```diff
@@ -40,17 +40,21 @@ class LoggingError(BaseException):
     pass
 
 
-class BaseModel(po.ConcreteModel):
```
[the remark this answers is not in the document]
Comment (p-snft) [reply src/oemof/solph/_models.py:43]: Indeed. There is only one model.

File: tests/test_solph_network_classes.py [hunk 2]
```diff
@@ -60,20 +61,22 @@ def test_sequence_conversion_factor_from_list_wrong_length(self):
...
-        assert trfr.inputs == {}
+        with pytest.warns(SuspiciousUsageWarning):
+            trfr = solph.components.Converter(outputs={})
+            assert trfr.inputs == {}
 
 
 def test_transformer_wrapper():
     with pytest.warns(FutureWarning):
-        solph.components.Transformer()
+        # no inputs/outputs
```
[the remark this answers is not in the document]
Comment (p-snft) [reply tests/test_solph_network_classes.py:77]: I wanted to explain why there are two `pytest.warns`.

File: tests/test_flows/test_non_convex_flow.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    assert (flow_result["flow"][:-1] == 3 * [5] + 7 * [0]).all()
+
+
+def test_activity_costs():
+    # activity costs higher then revenue for first time steps
+    flow = solph.flows.Flow(
+        nominal_value=10,
+        min=0.1,
+        max=[i * 0.1 for i in range(10)],
```
[the remark this answers is not in the document]
Comment (p-snft) [reply tests/test_flows/test_non_convex_flow.py:70]: This enforces the flow to be inactive. But you are right, this is not what we wanted. I'd suggest to have an input validation that `minimum <= maximum`.

File: tests/test_constraints/test_flow_count_limit.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    )
+    s1 = solph.components.Sink(
+        label="s2",
+        inputs={
+            b1: solph.Flow(
+                variable_costs=[-0.2, 0.2],
+                nominal_value=2,
+                min=0.5,
+                nonconvex=solph.NonConvex(),
+                custom_attributes={"keyword1": "also group 1"},
```
[the remark this answers is not in the document]
Comment (p-snft) [reply tests/test_constraints/test_flow_count_limit.py:45]: I want to make the point that the value is not relevant, here.

**Pull request merged**